### PR TITLE
Add MON_TYPES and MON_EGG_GROUPS

### LINKED ISSUE #3654: gSpeciesInfo could use a macro for types, egg groups and items to reduce redundancy
### Description

See #3562 

### Version

upcoming (Edge)

### Upcoming Version

_No response_

### Discord contact info

_No response_

### PR DESCRIPTION
**Pull request by Bassoonian**
Adds two metaprogram-based macros, MON_TYPES and MON_EGG_GROUPS, for use in gSpeciesInfo. These macros use the variable argument count to avoid having to specify the same type/egg group twice, which makes for easier and cleaner editing.

## Issue(s) that this PR fixes
Fixes #3654 

## **Discord contact info**
bassoonian

### CODE REVIEW

File: src/data/pokemon/species_info.h [hunk 1]
```diff
@@ -227,6 +227,9 @@ const u8 gOgerponCornerstoneMaskPokedexText[] = _(
 // 255 (MON_GENDERLESS) is reserved for genderless Pokémon.
 #define PERCENT_FEMALE(percent) min(254, ((percent * 255) / 100))
 
+#define MON_TYPE(type1, ...) { type1, DEFAULT(type1, __VA_ARGS__) }
+#define MON_EGG_GROUP(group1, ...) { group1, DEFAULT(group1, __VA_ARGS__) }
```
Comment (AsparagusEduardo) [inline src/data/pokemon/species_info.h:231]:
```suggestion
#define MON_TYPES(type1, ...) { type1, DEFAULT(type1, __VA_ARGS__) }
#define MON_EGG_GROUPS(group1, ...) { group1, DEFAULT(group1, __VA_ARGS__) }
```

**Comment by Bassoonian** [comment]
Ready for re-review

File: src/data/pokemon/species_info/gen_6.h [hunk 2]
```diff
@@ -408,7 +408,7 @@ const struct SpeciesInfo gSpeciesInfoGen6[] =
         ICON(Greninja, 0)
 
 #define GRENINJA_MISC_INFO                              \
-        .types = { TYPE_WATER, TYPE_DARK },             \
+        .types = MON_TYPES(TYPE_WATER, TYPE_DARK),       \
```
[the remark this answers is not in the document]
Comment (Bassoonian) [reply src/data/pokemon/species_info/gen_6.h:411]: Aw mannnn I manually fixed alignment everywhere and didn't realise that renaming them per your request completely ruined it, I'll take care of it later

[the remark this answers is not in the document]
Comment (AsparagusEduardo) [reply src/data/pokemon/species_info/gen_6.h:411]: Let me see if I can help with regex (I learned the hard way to not deal with mon data on-masse manually 🙌)

[the remark this answers is not in the document]
Comment (Bassoonian) [reply src/data/pokemon/species_info/gen_6.h]: Taken care of!

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**